### PR TITLE
qa: enable valgrind leak checks on ceph-mgr

### PR DESCRIPTION
**Pull request by liewegas**
Depends on https://github.com/ceph/teuthology/pull/1235 which suppresses all leaks in the Py code.  Clearly room for improvement there, but it was sufficient to find and fix several leaks in the rest of the mgr code!

There are several leak fixes we'll want to backport.  Opened http://tracker.ceph.com/issues/36744 to track the backports.

### CODE REVIEW

**Comment by tchaikov** [comment]
@liewegas could you rebase this PR? since the wip-mgr-commands has been merged.

**Comment by liewegas** [comment]
Maybe with @jcsp's py shutdown cleanup patch this won't show so many 
leaks..

**Comment by tchaikov** [comment]
@liewegas shall we fix all ceph-mgr valgrind issues before merging this PR or we can merge it first?

**Comment by liewegas** [comment]
We should fix them first or we'll get a ton of noise.  They may be 
resolved, though... it was a bunch of Py* stuff and teh shutdown and been 
changed a couple times since then.  I'll rebase and rerun a test.

**Comment by liewegas** [comment]
http://tracker.ceph.com/issues/19591 (i can take this one)
http://tracker.ceph.com/issues/19590 (Py stuff... want to look?)

**Comment by tchaikov** [comment]
will take care of http://tracker.ceph.com/issues/19590

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-04-21_02:04:57-rados-wip-sage-testing---basic-smithi/

still seeing leaks from python stuff:
<pre>
  <kind>Leak_PossiblyLost</kind>
  <xwhat>
    <text>38 bytes in 1 blocks are possibly lost in loss record 623 of 4,188</text>
    <leakedbytes>38</leakedbytes>
    <leakedblocks>1</leakedblocks>
  </xwhat>
  <stack>
    <frame>
      <ip>0x981CBE3</ip>
      <obj>/usr/lib64/valgrind/vgpreload_memcheck-amd64-linux.so</obj>
      <fn>malloc</fn>
      <dir>/builddir/build/BUILD/valgrind-3.11.0/coregrind/m_replacemalloc</dir>
      <file>vg_replace_malloc.c</file>
      <line>299</line>
    </frame>
    <frame>
      <ip>0x9AB1CCC</ip>
      <obj>/usr/lib64/libpython2.7.so.1.0</obj>
      <fn>PyObject_Malloc</fn>
    </frame>
    <frame>
      <ip>0x9AC1974</ip>
      <obj>/usr/lib64/libpython2.7.so.1.0</obj>
      <fn>PyString_FromString</fn>
    </frame>
    <frame>
      <ip>0x9A9020F</ip>
      <obj>/usr/lib64/libpython2.7.so.1.0</obj>
    </frame>
    <frame>
      <ip>0x9A90B07</ip>
      <obj>/usr/lib64/libpython2.7.so.1.0</obj>
      <fn>PyFile_FromFile</fn>
    </frame>
</pre>

**Comment by tchaikov** [comment]
the run with valgrind is http://pulpito.ceph.com/sage-2017-04-21_02:04:57-rados-wip-sage-testing---basic-smithi/1050599

**Comment by tchaikov** [comment]
created http://tracker.ceph.com/issues/19780 to track the related efforts.

**Comment by liewegas** [comment]
@tchaikov i can't remember where we ended up on this...

**Comment by tchaikov** [comment]
@liewegas i added a bunch suppressions. will give it another try.

**Comment by tchaikov** [comment]
/a/kchai-2018-01-08_11:42:05-rados-wip-kefu-testing-2018-01-08-1156-distro-basic-mira/2042713, there are still a couple of leaks identified by valgrind.

**Comment by jcsp** [comment]
Some more updates to valgrind.supp:
https://github.com/ceph/teuthology/pull/1145

**Comment by liewegas** [comment]
will give this another go.. worst case, we whitelist all of the python allocations?

File: src/mon/MonClient.h [hunk 1]
```diff
@@ -200,7 +201,7 @@ class MonClient : public Dispatcher {
   int authenticate_err = 0;
   bool authenticated = false;
 
-  list<Message*> waiting_for_session;
+  list<MessageRef> waiting_for_session;
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/mon/MonClient.h]: agree, i misunderstood the bug originally :)

**Comment by liewegas** [comment]
Woohoo!  http://pulpito.ceph.com/sage-2018-11-16_09:33:08-rados:verify-wip-sage3-testing-2018-11-15-2334-distro-basic-smithi/

This needs to merge with #25052 tho; testing them together now.

**Comment by liewegas** [comment]
retest this please

**Pull request merged**